### PR TITLE
Ensure only the pin code is sent

### PR DESCRIPTION
**Pull request by emelenas**
The pinentry program answers with "D PIN\nOK\n". Sometimes the "\nOK\n" gets sent to the gateway, causing errors. The change ensures only the PIN is sent, by taking the string up to, and discarding, the first "\n". See https://github.com/adrienverge/openfortivpn/issues/744

### CODE REVIEW

File: src/userinput.c [hunk 1]
```diff
@@ -135,6 +135,7 @@ static int pinentry_read(int from, char **retstr)
 	    || strncmp(buf, "D ", 2) == 0) {
 		if (retstr) {
 			*retstr = strchr(buf, ' ');
+			*retstr = strtok(*retstr, "\n");
```
Comment (DimitriPapadopoulos) [inline src/userinput.c:138]: Use **`strtok_r`** instead of `strtok`. See #571.

Comment (emelenas) [reply src/userinput.c:138]:
```suggestion
			*retstr = strtok_r(*retstr, "\n");
```

Comment (emelenas) [reply src/userinput.c:138]: O, I was not aware

Comment (DimitriPapadopoulos) [reply src/userinput.c:138]: [Function `strtok_r` requires 3 arguments](https://man7.org/linux/man-pages/man3/strtok_r.3.html).

Comment (emelenas) [reply src/userinput.c:138]: OK, sorry. I didn't test.

```suggestion
                        char *saveptr = NULL;
			*retstr = *retstr ? strtok_r(*retstr, "\n", &saveptr) : NULL;
```

Comment (emelenas) [reply src/userinput.c:138]: I'll update the fork and make a new merge request.

File: src/userinput.c [hunk 2]
```diff
@@ -135,6 +135,7 @@ static int pinentry_read(int from, char **retstr)
 	    || strncmp(buf, "D ", 2) == 0) {
 		if (retstr) {
 			*retstr = strchr(buf, ' ');
+			**retstr = *retstr ? strtok_r(*retstr, "\n", &saveptr) : NULL;
```
Comment (DimitriPapadopoulos) [inline src/userinput.c:138]: For a single call to `strtok_r` I believe you don't need `saveptr`, isn't `NULL` enough as the 3rd argument?

Also this is suspect:
```
**retstr = *retstr [...]
```
Perhaps is should be:
```
*retstr = *retstr [...]
```

**Comment by emelenas** [comment]
Maybe I get it right this time? Sorry, I shouldn't be multitasking.

**Comment by DimitriPapadopoulos** [comment]
Looks good to me now. Have or could you build and install this version of _openfortivpn_ and make sure it works with _NetworkManager_ and fixes the reported issue?

**Comment by emelenas** [comment]
I have built it and tested OK, but I will ask the original poster to also test it

File: src/userinput.c [hunk 3]
```diff
@@ -135,6 +135,7 @@ static int pinentry_read(int from, char **retstr)
 	    || strncmp(buf, "D ", 2) == 0) {
 		if (retstr) {
 			*retstr = strchr(buf, ' ');
+			*retstr = *retstr ? strtok_r(*retstr, "\n", &saveptr) : NULL;
```
[the remark this answers is not in the document]
Comment (emelenas) [reply src/userinput.c:138]: Sorry, it does not compile with NULL as argument:
src/userinput.c:138:24: warning: null argument where non-null required (argument 3) [-Wnonnull]

[the remark this answers is not in the document]
Comment (DimitriPapadopoulos) [reply src/userinput.c:138]: Ah, you're absolutely right. Besides we use `saveptr` everywhere else.

**Pull request merged**